File: solutions/haiku-1/haiku/go/dijkstra.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

// Item represents a node with its distance in the priority queue
type Item struct {
	node     int
	distance int
	index    int
}

// PriorityQueue implements heap.Interface
type PriorityQueue []*Item

func (pq PriorityQueue) Len() int { return len(pq) }
func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].distance < pq[j].distance
}
func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *PriorityQueue) Push(x interface{}) {
	n := len(*pq)
	item := x.(*Item)
	item.index = n
	*pq = append(*pq, item)
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil
	item.index = -1
	*pq = old[0 : n-1]
	return item
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	// Read N and M
	scanner.Scan()
	parts := strings.Fields(scanner.Text())
	n, _ := strconv.Atoi(parts[0])
	m, _ := strconv.Atoi(parts[1])

	// Build adjacency list
	graph := make([][]struct {
		node   int
		weight int
	}, n)

	for i := 0; i < m; i++ {
		scanner.Scan()
		parts := strings.Fields(scanner.Text())
		u, _ := strconv.Atoi(parts[0])
		v, _ := strconv.Atoi(parts[1])
		w, _ := strconv.Atoi(parts[2])

		graph[u] = append(graph[u], struct {
			node   int
			weight int
		}{v, w})
		graph[v] = append(graph[v], struct {
			node   int
			weight int
		}{u, w})
	}

	// Read start and target
	scanner.Scan()
	parts = strings.Fields(scanner.Text())
	s, _ := strconv.Atoi(parts[0])
	t, _ := strconv.Atoi(parts[1])

	// Dijkstra's algorithm with priority queue
	dist := make([]int, n)
	for i := range dist {
		dist[i] = math.MaxInt
	}
	dist[s] = 0

	pq := make(PriorityQueue, 0)
	heap.Push(&pq, &Item{node: s, distance: 0})

	for pq.Len() > 0 {
		current := heap.Pop(&pq).(*Item)
		u := current.node

		// If we've already found a shorter path to this node, skip
		if current.distance > dist[u] {
			continue
		}

		// If we've reached the target, we can return early
		if u == t {
			break
		}

		// Explore neighbors
		for _, edge := range graph[u] {
			v := edge.node
			w := edge.weight
			newDist := dist[u] + w

			if newDist < dist[v] {
				dist[v] = newDist
				heap.Push(&pq, &Item{node: v, distance: newDist})
			}
		}
	}

	if dist[t] == math.MaxInt {
		fmt.Println(-1)
	} else {
		fmt.Println(dist[t])
	}
}

File: solutions/haiku-1/haiku/rust/dijkstra.rs
use std::collections::BinaryHeap;
use std::cmp::Ordering;
use std::io::{self, BufRead};

#[derive(Eq, PartialEq)]
struct State {
    cost: usize,
    node: usize,
}

impl Ord for State {
    fn cmp(&self, other: &Self) -> Ordering {
        // Reverse comparison for min-heap
        other.cost.cmp(&self.cost)
            .then_with(|| self.node.cmp(&other.node))
    }
}

impl PartialOrd for State {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    // Read N and M
    let first_line = lines.next().unwrap().unwrap();
    let parts: Vec<&str> = first_line.split_whitespace().collect();
    let n: usize = parts[0].parse().unwrap();
    let m: usize = parts[1].parse().unwrap();

    // Build adjacency list
    let mut graph: Vec<Vec<(usize, usize)>> = vec![Vec::new(); n];

    for _ in 0..m {
        let line = lines.next().unwrap().unwrap();
        let parts: Vec<&str> = line.split_whitespace().collect();
        let u: usize = parts[0].parse().unwrap();
        let v: usize = parts[1].parse().unwrap();
        let w: usize = parts[2].parse().unwrap();

        graph[u].push((v, w));
        graph[v].push((u, w));
    }

    // Read s and t
    let last_line = lines.next().unwrap().unwrap();
    let parts: Vec<&str> = last_line.split_whitespace().collect();
    let s: usize = parts[0].parse().unwrap();
    let t: usize = parts[1].parse().unwrap();

    // Handle special case where s equals t
    if s == t {
        println!("0");
        return;
    }

    // Dijkstra's algorithm
    let mut distances = vec![usize::MAX; n];
    let mut heap = BinaryHeap::new();

    distances[s] = 0;
    heap.push(State { cost: 0, node: s });

    while let Some(State { cost, node }) = heap.pop() {
        // If we've already found a shorter path, skip this state
        if cost > distances[node] {
            continue;
        }

        // If we reached the target, output and exit
        if node == t {
            println!("{}", cost);
            return;
        }

        // Explore neighbors
        for &(next_node, weight) in &graph[node] {
            let next_cost = cost + weight;

            if next_cost < distances[next_node] {
                distances[next_node] = next_cost;
                heap.push(State { cost: next_cost, node: next_node });
            }
        }
    }

    // No path found
    println!("-1");
}
